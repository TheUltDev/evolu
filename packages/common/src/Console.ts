/**
 * üìù Cross-Platform Console
 *
 * Console abstraction for Chrome 123+, Firefox 125+, Safari 18.1+, Node.js
 * 22.x+, and React Native 0.75+. Includes methods guaranteed to be available in
 * these environments and expected to remain compatible in future versions.
 * Output formatting may vary (e.g., interactive UI in browsers vs. text in
 * Node.js/React Native), but functionality is consistent across platforms.
 *
 * **Warning**: If you encounter platform-specific issues or missing methods,
 * please contribute a PR with details about the environment and behavior.
 *
 * @module
 */

/**
 * Cross-platform Console interface for Chrome 123+, Firefox 125+, Safari 18.1+,
 * Node.js 22.x+, and React Native 0.75+
 */
export interface Console {
  /** Controls whether console methods produce output (default: true) */
  enabled?: boolean;

  /** Outputs a message to the console */
  log: (...args: Array<any>) => void;

  /** Outputs an informational message (often same as log) */
  info: (...args: Array<any>) => void;

  /** Outputs a warning message */
  warn: (...args: Array<any>) => void;

  /** Outputs an error message */
  error: (...args: Array<any>) => void;

  /** Outputs a debug message */
  debug: (...args: Array<any>) => void;

  /** Starts a timer with an optional label */
  time: (label?: string) => void;

  /** Ends a timer and logs the elapsed time */
  timeEnd: (label?: string) => void;

  /** Displays an object's properties in a detailed format */
  dir: (object: any, options?: any) => void;

  /** Displays tabular data as a table */
  table: (tabularData: any, properties?: Array<string>) => void;

  /** Logs the number of times this has been called with the given label */
  count: (label?: string) => void;

  /** Resets the counter for the given label */
  countReset: (label?: string) => void;

  /** Writes a message if the value is falsy, otherwise does nothing */
  assert: (value: any, message?: string, ...optionalParams: Array<any>) => void;

  /** Logs the elapsed time for a timer without ending it */
  timeLog: (label?: string, ...data: Array<any>) => void;

  /** Prints a stack trace with an optional message */
  trace: (message?: any, ...optionalParams: Array<any>) => void;
}

/** Dependency interface for injecting a Console instance. */
export interface ConsoleDep {
  readonly console: Console;
}

export interface ConsoleConfig {
  /**
   * Enable or disable console logging. When true, logs are output to the
   * {@link Console}; when false, logging is disabled for all methods except
   * `error`, which always outputs to ensure critical issues are not missed.
   * Default is false.
   */
  readonly enableLogging?: boolean;
}

/** Creates a console instance using the global console. */
export const createConsole = (config: ConsoleConfig = {}): Console => {
  const instance: Console = {
    enabled: config.enableLogging ?? false,

    log: (...args) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.log(...args);
    },
    info: (...args) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.info(...args);
    },
    warn: (...args) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.warn(...args);
    },
    error: (...args) => {
      // Always log errors, even if disabled
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      console.error(...args);
    },
    debug: (...args) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.debug(...args);
    },
    time: (label) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.time(label);
    },
    timeEnd: (label) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.timeEnd(label);
    },
    dir: (object, options) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.dir(object, options);
    },
    table: (tabularData, properties) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.table(tabularData, properties);
    },
    count: (label) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.count(label);
    },
    countReset: (label) => {
      // eslint-disable-next-line no-console
      if (instance.enabled) console.countReset(label);
    },
    assert: (value, message, ...optionalParams) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.assert(value, message, ...optionalParams);
    },
    timeLog: (label, ...data) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.timeLog(label, ...data);
    },
    trace: (message, ...optionalParams) => {
      // eslint-disable-next-line no-console, @typescript-eslint/no-unsafe-argument
      if (instance.enabled) console.trace(message, ...optionalParams);
    },
  };

  return instance;
};
