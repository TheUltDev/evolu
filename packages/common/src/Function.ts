/**
 * Helper function to ensure exhaustive matching in a switch statement. Throws
 * an error if an unhandled case is encountered.
 *
 * Remember, it's useful only when we don't return anything from the switch
 * statement. Otherwise, a return type of a function is enough.
 *
 * ### Example
 *
 * ```ts
 * type Color = "red" | "green" | "blue";
 *
 * function handleColor(color: Color): void {
 *   switch (color) {
 *     case "red":
 *       console.log("Handling red");
 *       break;
 *     case "green":
 *       console.log("Handling green");
 *       break;
 *     case "blue":
 *       console.log("Handling blue");
 *       break;
 *     default:
 *       exhaustiveCheck(color); // Ensures all cases are handled
 *   }
 * }
 * ```
 */
export const exhaustiveCheck = (value: never): never => {
  throw new Error(`exhaustiveCheck unhandled case: ${JSON.stringify(value)}`);
};

export const identity = <A>(a: A): A => a;

/**
 * A function that delays computation and returns a value of type T.
 *
 * Useful for:
 *
 * - Lazy evaluation
 * - Returning constant values
 * - Providing default or placeholder behaviors
 *
 * ### Example
 *
 * ```ts
 * const getRandomNumber: LazyValue<number> = () => Math.random();
 * const randomValue = getRandomNumber();
 * ```
 */
export type LazyValue<T> = () => T;

export const constVoid: LazyValue<void> = () => undefined;
export const constUndefined: LazyValue<undefined> = () => undefined;
export const constNull: LazyValue<null> = () => null;
export const constTrue: LazyValue<true> = () => true;
export const constFalse: LazyValue<false> = () => false;
