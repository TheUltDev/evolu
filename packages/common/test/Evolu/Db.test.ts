import { assert, expect, test } from "vitest";
import { CallbackId } from "../../src/Callbacks.js";
import { Console } from "../../src/Console.js";
import {
  createDbWorkerForPlatform,
  DbSchema,
  DbSnapshot,
  DbWorkerOutput,
  DbWorkerPlatformDeps,
  getDbSchema,
  getDbSnapshot,
  maybeMigrateToVersion0,
} from "../../src/Evolu/Db.js";
import { ColumnName, TableName } from "../../src/Evolu/Protocol.js";
import { defaultColumnsNames } from "../../src/Evolu/Schema.js";
import { constVoid } from "../../src/Function.js";
import { wait } from "../../src/Promise.js";
import { createRandomWithSeed } from "../../src/Random.js";
import { getOrThrow } from "../../src/Result.js";
import { createSqlite, sql, Sqlite } from "../../src/Sqlite.js";
import { createId } from "../../src/Type.js";
import {
  testCreateId,
  testCreateMnemonic,
  testCreateRandomBytesDep,
  testCreateSqlite,
  testCreateSqliteDriver,
  testDbConfig,
  testNanoIdLib,
  testNanoIdLibDep,
  testSimpleName,
  testTime,
} from "../_deps.js";

export const testCreateVersion0 = ({ exec }: Sqlite): void => {
  exec(sql`begin;`);

  exec(sql`
    create table "evolu_message" (
      "timestamp" blob primary key,
      "table" blob,
      "row" blob,
      "column" blob,
      "value" blob
    );
  `);

  exec(sql`
    create index "index_evolu_message" on "evolu_message" (
      "table",
      "row",
      "column",
      "timestamp" desc
    );
  `);

  exec(sql`
    insert into evolu_message
    values
      (
        '2024-08-24T07:16:20.246Z-0000-431898faea378430',
        'todoCategory',
        'VeLlYHBvFcy3yAhboJp-L',
        'name',
        'Not Urgent'
      );
  `);

  exec(sql`
    insert into evolu_message
    values
      (
        '2024-08-24T07:16:20.246Z-0001-431898faea378430',
        'todo',
        'JiSMETkhgJT-GZ_2W3p48',
        'title',
        'Try React Suspense'
      );
  `);

  exec(sql`
    insert into evolu_message
    values
      (
        '2024-08-24T07:16:20.246Z-0002-431898faea378430',
        'todo',
        'JiSMETkhgJT-GZ_2W3p48',
        'categoryId',
        'VeLlYHBvFcy3yAhboJp-L'
      );
  `);

  exec(sql`
    insert into evolu_message
    values
      (
        '2024-08-24T07:16:28.969Z-0000-431898faea378430',
        'todo',
        '18DyYqVEmVI0voIlxbniY',
        'title',
        'Test 1'
      );
  `);

  exec(sql`
    insert into evolu_message
    values
      (
        '2024-09-05T12:54:42.857Z-0000-431898faea378430',
        'todo',
        'ZMmNCbACx5Km29e5wZ48l',
        'isCompleted',
        1
      );
  `);

  exec(sql`
    create table "evolu_owner" (
      "id" blob,
      "mnemonic" blob,
      "encryptionkey" blob,
      "timestamp" blob,
      "merkletree" blob
    );
  `);

  exec(sql`
    insert into evolu_owner
    values
      (
        'NXqRROM1U3dSJRawS9LRf',
        'over elder sense peace scheme hard total pigeon access tomato spray ocean',
        x'8d6b3447d9a6ae6890eead6b713b1c297e7af26ec7eb4cb5d90f5657d9091539',
        '2024-09-05T14:07:03.338Z-0004-de562c45893915d4',
        '{"2":{"0":{"0":{"0":{"0":{"0":{"2":{"0":{"1":{"2":{"2":{"1":{"0":{"0":{"1":{"1":{"hash":676696346},"hash":676696346},"hash":676696346},"hash":676696346},"hash":676696346},"hash":676696346},"hash":676696346},"hash":676696346},"hash":676696346},"hash":676696346},"hash":676696346},"1":{"0":{"0":{"0":{"0":{"2":{"1":{"2":{"1":{"0":{"1":{"hash":1532565686},"hash":1532565686},"hash":1532565686},"hash":1532565686},"hash":1532565686},"hash":1532565686},"hash":1532565686},"2":{"2":{"2":{"1":{"1":{"1":{"0":{"hash":1149993009},"1":{"hash":1677793058},"2":{"hash":-1929932990},"hash":-1401087919},"2":{"0":{"hash":-1331646170},"1":{"hash":1147271680},"2":{"hash":1003866511},"hash":-820739415},"hash":1667865336},"2":{"0":{"0":{"hash":-1060739940},"hash":-1060739940},"hash":-1060739940},"hash":-1548747164},"hash":-1548747164},"hash":-1548747164},"hash":-1548747164},"hash":-118036782},"hash":-118036782},"hash":-118036782},"hash":-118036782},"hash":-794593336},"hash":-794593336},"hash":-794593336},"hash":-794593336},"hash":-794593336},"hash":-794593336}'
      );
  `);

  exec(sql`commit;`);
};

// TODO: Migration works, but we have to update the test.
test.skip("maybeMigrateTo0", async () => {
  const sqlite = await testCreateSqlite();
  testCreateVersion0(sqlite);

  const schema = getDbSchema({ sqlite })();
  assert(schema.ok);

  maybeMigrateToVersion0({
    ...testCreateRandomBytesDep,
    sqlite,
  })(schema.value);

  const snapshot = getDbSnapshot({ sqlite });

  expect(snapshot).toMatchSnapshot();
});

test("createDbWorker", async () => {
  const onMessageMessagesAndDbSnapshots: Array<{
    message: DbWorkerOutput;
    snapshot: DbSnapshot;
  }> = [];
  const logArgs: Array<unknown> = [];

  const sqliteDriver = await testCreateSqliteDriver(testSimpleName);
  const createSqliteDriver = () => Promise.resolve(sqliteDriver);
  const sqlite = getOrThrow(
    await createSqlite({ createSqliteDriver })(testSimpleName),
  );

  const deps: DbWorkerPlatformDeps = {
    createSqliteDriver,
    createSync: () => () => ({
      send: constVoid,
    }),
    console: {
      log: (...args) => {
        logArgs.push(args);
      },
    } as Console,
    time: testTime,
    random: createRandomWithSeed("test"),
    nanoIdLib: testNanoIdLib,
    createMnemonic: testCreateMnemonic,
    ...testCreateRandomBytesDep,
  };

  const db = createDbWorkerForPlatform(deps);

  db.onMessage((message) => {
    const snapshot = getDbSnapshot({ sqlite });
    assert(snapshot.ok);
    onMessageMessagesAndDbSnapshots.push({
      message,
      snapshot: snapshot.value,
    });
  });

  const tabId = createId(testNanoIdLibDep);

  // Run mutate before the init to test the init is handled first and mutate waits.
  db.postMessage({
    type: "mutate",
    tabId,
    changes: [
      {
        id: testCreateId(),
        table: "_table1" as TableName,
        values: {
          ["column1" as ColumnName]: "bar",
        },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  const testDbSchema: DbSchema = {
    tables: [
      {
        name: "_table1" as TableName,
        columns: ["id" as ColumnName, "column1" as ColumnName].concat(
          defaultColumnsNames,
        ),
      },
      {
        name: "table1" as TableName,
        columns: ["id" as ColumnName, "column1" as ColumnName].concat(
          defaultColumnsNames,
        ),
      },
    ],
    indexes: [],
  };

  db.postMessage({
    type: "init",
    config: { ...testDbConfig, enableLogging: true },
    dbSchema: testDbSchema,
    initialData: [
      {
        id: testCreateId(),
        table: "table1" as TableName,
        values: {
          ["column1" as ColumnName]: "foo",
        },
      },
    ],
  });

  db.postMessage({
    type: "mutate",
    tabId,
    changes: [
      {
        id: testCreateId(),
        table: "_table1" as TableName,
        values: {
          ["isDeleted" as ColumnName]: 1,
        },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  db.postMessage({
    type: "mutate",
    tabId,
    changes: [
      {
        id: testCreateId(),
        table: "Table1" as TableName,
        values: {
          ["column1" as ColumnName]: "foo",
        },
      },
    ],
    onCompleteIds: [],
    subscribedQueries: [],
  });

  db.postMessage({
    type: "reset",
    reload: false,
    onCompleteId: testNanoIdLib.nanoid() as CallbackId,
  });

  await wait(10);

  expect(logArgs).toMatchSnapshot("logArgs");

  expect(onMessageMessagesAndDbSnapshots).toMatchSnapshot(
    "onMessageMessagesAndDbSnapshots",
  );

  // TODO:
  // db.ensureDbSchema({ tables: [], indexes: [] });
  // getDatabaseSnapshot.toMatchSnapshot(
  //   "ensureDbSchema doesn't remove tables nor app indexes",
  // );
  // const indexes = createIndexes((create) => [
  //   create("indexTodoCreatedAt").on("todo").column("createdAt"),
  // ]);
  // db.ensureDbSchema({ tables: [], indexes })
  // getDatabaseSnapshot.toMatchSnapshot(
  //   "ensureDbSchema adds user indexes",
  // );
});
