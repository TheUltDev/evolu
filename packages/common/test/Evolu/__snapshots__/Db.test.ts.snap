// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createDbWorker initializes correctly 1`] = `
[
  {
    "appOwner": {
      "encryptionKey": Uint8Array [
        214,
        142,
        243,
        254,
        221,
        13,
        246,
        19,
        136,
        198,
        193,
        14,
        183,
        237,
        206,
        51,
        171,
        193,
        65,
        231,
        26,
        97,
        31,
        142,
        157,
        180,
        17,
        113,
        106,
        171,
        129,
        26,
      ],
      "id": "B1BmuTBy1NJ4LZLWnfokr",
      "mnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
      "type": "AppOwner",
      "writeKey": Uint8Array [
        193,
        200,
        155,
        104,
        173,
        156,
        172,
        55,
        27,
        230,
        228,
        51,
        167,
        204,
        43,
        8,
      ],
    },
    "isFirst": true,
    "type": "onInit",
  },
]
`;

exports[`createDbWorker initializes correctly 2`] = `
{
  "schema": {
    "indexes": [
      {
        "name": "evolu_history_ownerId_timestamp",
        "sql": "create index evolu_history_ownerId_timestamp on evolu_history (
          "ownerId",
          "timestamp"
        )",
      },
      {
        "name": "evolu_history_ownerId_table_id_column_timestampDesc",
        "sql": "create unique index evolu_history_ownerId_table_id_column_timestampDesc on evolu_history (
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp" desc
        )",
      },
      {
        "name": "evolu_timestamp_index",
        "sql": "create index evolu_timestamp_index on evolu_timestamp (
        "ownerId",
        "l",
        "t",
        "h1",
        "h2",
        "c"
      )",
      },
    ],
    "tables": [
      {
        "columns": [
          "protocolVersion",
        ],
        "name": "evolu_version",
      },
      {
        "columns": [
          "clock",
          "appOwnerId",
          "appOwnerEncryptionKey",
          "appOwnerWriteKey",
          "appOwnerMnemonic",
        ],
        "name": "evolu_config",
      },
      {
        "columns": [
          "ownerId",
          "table",
          "id",
          "column",
          "timestamp",
          "value",
        ],
        "name": "evolu_history",
      },
      {
        "columns": [
          "id",
          "name",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "testTable",
      },
      {
        "columns": [
          "id",
          "name",
          "createdAt",
          "updatedAt",
          "isDeleted",
        ],
        "name": "_testTable",
      },
      {
        "columns": [
          "ownerId",
          "t",
          "h1",
          "h2",
          "c",
          "l",
        ],
        "name": "evolu_timestamp",
      },
    ],
  },
  "tables": [
    {
      "name": "evolu_version",
      "rows": [
        {
          "protocolVersion": 0,
        },
      ],
    },
    {
      "name": "evolu_config",
      "rows": [
        {
          "appOwnerEncryptionKey": {
            "data": [
              214,
              142,
              243,
              254,
              221,
              13,
              246,
              19,
              136,
              198,
              193,
              14,
              183,
              237,
              206,
              51,
              171,
              193,
              65,
              231,
              26,
              97,
              31,
              142,
              157,
              180,
              17,
              113,
              106,
              171,
              129,
              26,
            ],
            "type": "Buffer",
          },
          "appOwnerId": "B1BmuTBy1NJ4LZLWnfokr",
          "appOwnerMnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
          "appOwnerWriteKey": {
            "data": [
              193,
              200,
              155,
              104,
              173,
              156,
              172,
              55,
              27,
              230,
              228,
              51,
              167,
              204,
              43,
              8,
            ],
            "type": "Buffer",
          },
          "clock": "1970-01-01T00:00:00.000Z-0000-0fb0b7912c7c1fca",
        },
      ],
    },
    {
      "name": "evolu_history",
      "rows": [],
    },
    {
      "name": "testTable",
      "rows": [],
    },
    {
      "name": "_testTable",
      "rows": [],
    },
    {
      "name": "evolu_timestamp",
      "rows": [],
    },
  ],
}
`;

exports[`local mutation 1`] = `
[
  {
    "appOwner": {
      "encryptionKey": Uint8Array [
        214,
        142,
        243,
        254,
        221,
        13,
        246,
        19,
        136,
        198,
        193,
        14,
        183,
        237,
        206,
        51,
        171,
        193,
        65,
        231,
        26,
        97,
        31,
        142,
        157,
        180,
        17,
        113,
        106,
        171,
        129,
        26,
      ],
      "id": "B1BmuTBy1NJ4LZLWnfokr",
      "mnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
      "type": "AppOwner",
      "writeKey": Uint8Array [
        193,
        200,
        155,
        104,
        173,
        156,
        172,
        55,
        27,
        230,
        228,
        51,
        167,
        204,
        43,
        8,
      ],
    },
    "isFirst": true,
    "type": "onInit",
  },
  {
    "onCompleteIds": [],
    "patches": [],
    "tabId": "AAUWg7IbJo2XpJicKJoim",
    "type": "onChange",
  },
  {
    "tabId": "AAUWg7IbJo2XpJicKJoim",
    "type": "onReceive",
  },
]
`;

exports[`local mutation 2`] = `
[
  {
    "name": "evolu_version",
    "rows": [
      {
        "protocolVersion": 0,
      },
    ],
  },
  {
    "name": "evolu_config",
    "rows": [
      {
        "appOwnerEncryptionKey": {
          "data": [
            214,
            142,
            243,
            254,
            221,
            13,
            246,
            19,
            136,
            198,
            193,
            14,
            183,
            237,
            206,
            51,
            171,
            193,
            65,
            231,
            26,
            97,
            31,
            142,
            157,
            180,
            17,
            113,
            106,
            171,
            129,
            26,
          ],
          "type": "Buffer",
        },
        "appOwnerId": "B1BmuTBy1NJ4LZLWnfokr",
        "appOwnerMnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
        "appOwnerWriteKey": {
          "data": [
            193,
            200,
            155,
            104,
            173,
            156,
            172,
            55,
            27,
            230,
            228,
            51,
            167,
            204,
            43,
            8,
          ],
          "type": "Buffer",
        },
        "clock": "1970-01-01T00:00:00.000Z-0000-0db883b25f327a10",
      },
    ],
  },
  {
    "name": "evolu_history",
    "rows": [],
  },
  {
    "name": "testTable",
    "rows": [],
  },
  {
    "name": "_testTable",
    "rows": [
      {
        "createdAt": "1970-01-01T00:00:00.002Z",
        "id": "E7z9KY-wzYDMkAArIfacN",
        "isDeleted": null,
        "name": "test",
        "updatedAt": "1970-01-01T00:00:00.002Z",
      },
    ],
  },
  {
    "name": "evolu_timestamp",
    "rows": [],
  },
]
`;

exports[`local mutation 3`] = `
[
  {
    "name": "evolu_version",
    "rows": [
      {
        "protocolVersion": 0,
      },
    ],
  },
  {
    "name": "evolu_config",
    "rows": [
      {
        "appOwnerEncryptionKey": {
          "data": [
            214,
            142,
            243,
            254,
            221,
            13,
            246,
            19,
            136,
            198,
            193,
            14,
            183,
            237,
            206,
            51,
            171,
            193,
            65,
            231,
            26,
            97,
            31,
            142,
            157,
            180,
            17,
            113,
            106,
            171,
            129,
            26,
          ],
          "type": "Buffer",
        },
        "appOwnerId": "B1BmuTBy1NJ4LZLWnfokr",
        "appOwnerMnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
        "appOwnerWriteKey": {
          "data": [
            193,
            200,
            155,
            104,
            173,
            156,
            172,
            55,
            27,
            230,
            228,
            51,
            167,
            204,
            43,
            8,
          ],
          "type": "Buffer",
        },
        "clock": "1970-01-01T00:00:00.000Z-0000-0db883b25f327a10",
      },
    ],
  },
  {
    "name": "evolu_history",
    "rows": [],
  },
  {
    "name": "testTable",
    "rows": [],
  },
  {
    "name": "_testTable",
    "rows": [],
  },
  {
    "name": "evolu_timestamp",
    "rows": [],
  },
]
`;

exports[`mutate before init 1`] = `
[
  {
    "appOwner": {
      "encryptionKey": Uint8Array [
        214,
        142,
        243,
        254,
        221,
        13,
        246,
        19,
        136,
        198,
        193,
        14,
        183,
        237,
        206,
        51,
        171,
        193,
        65,
        231,
        26,
        97,
        31,
        142,
        157,
        180,
        17,
        113,
        106,
        171,
        129,
        26,
      ],
      "id": "B1BmuTBy1NJ4LZLWnfokr",
      "mnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
      "type": "AppOwner",
      "writeKey": Uint8Array [
        193,
        200,
        155,
        104,
        173,
        156,
        172,
        55,
        27,
        230,
        228,
        51,
        167,
        204,
        43,
        8,
      ],
    },
    "isFirst": true,
    "type": "onInit",
  },
  {
    "onCompleteIds": [],
    "patches": [],
    "tabId": "APRJEKIftYOy3PChCWos2",
    "type": "onChange",
  },
  {
    "tabId": "APRJEKIftYOy3PChCWos2",
    "type": "onReceive",
  },
]
`;

exports[`mutate before init 2`] = `
[
  {
    "name": "evolu_version",
    "rows": [
      {
        "protocolVersion": 0,
      },
    ],
  },
  {
    "name": "evolu_config",
    "rows": [
      {
        "appOwnerEncryptionKey": {
          "data": [
            214,
            142,
            243,
            254,
            221,
            13,
            246,
            19,
            136,
            198,
            193,
            14,
            183,
            237,
            206,
            51,
            171,
            193,
            65,
            231,
            26,
            97,
            31,
            142,
            157,
            180,
            17,
            113,
            106,
            171,
            129,
            26,
          ],
          "type": "Buffer",
        },
        "appOwnerId": "B1BmuTBy1NJ4LZLWnfokr",
        "appOwnerMnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
        "appOwnerWriteKey": {
          "data": [
            193,
            200,
            155,
            104,
            173,
            156,
            172,
            55,
            27,
            230,
            228,
            51,
            167,
            204,
            43,
            8,
          ],
          "type": "Buffer",
        },
        "clock": "1970-01-01T00:00:00.000Z-0000-b8442d71bc503af0",
      },
    ],
  },
  {
    "name": "evolu_history",
    "rows": [],
  },
  {
    "name": "testTable",
    "rows": [],
  },
  {
    "name": "_testTable",
    "rows": [
      {
        "createdAt": "1970-01-01T00:00:00.001Z",
        "id": "_dGi2jCaTw-bM9tdhdvUI",
        "isDeleted": null,
        "name": "test",
        "updatedAt": "1970-01-01T00:00:00.001Z",
      },
    ],
  },
  {
    "name": "evolu_timestamp",
    "rows": [],
  },
]
`;

exports[`mutations 1`] = `
[
  {
    "appOwner": {
      "encryptionKey": Uint8Array [
        214,
        142,
        243,
        254,
        221,
        13,
        246,
        19,
        136,
        198,
        193,
        14,
        183,
        237,
        206,
        51,
        171,
        193,
        65,
        231,
        26,
        97,
        31,
        142,
        157,
        180,
        17,
        113,
        106,
        171,
        129,
        26,
      ],
      "id": "B1BmuTBy1NJ4LZLWnfokr",
      "mnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
      "type": "AppOwner",
      "writeKey": Uint8Array [
        193,
        200,
        155,
        104,
        173,
        156,
        172,
        55,
        27,
        230,
        228,
        51,
        167,
        204,
        43,
        8,
      ],
    },
    "isFirst": true,
    "type": "onInit",
  },
  {
    "onCompleteIds": [],
    "patches": [],
    "tabId": "4W_18Oy2YBz9_EC0v62hC",
    "type": "onChange",
  },
  {
    "tabId": "4W_18Oy2YBz9_EC0v62hC",
    "type": "onReceive",
  },
]
`;

exports[`mutations 2`] = `
[
  {
    "name": "evolu_version",
    "rows": [
      {
        "protocolVersion": 0,
      },
    ],
  },
  {
    "name": "evolu_config",
    "rows": [
      {
        "appOwnerEncryptionKey": {
          "data": [
            214,
            142,
            243,
            254,
            221,
            13,
            246,
            19,
            136,
            198,
            193,
            14,
            183,
            237,
            206,
            51,
            171,
            193,
            65,
            231,
            26,
            97,
            31,
            142,
            157,
            180,
            17,
            113,
            106,
            171,
            129,
            26,
          ],
          "type": "Buffer",
        },
        "appOwnerId": "B1BmuTBy1NJ4LZLWnfokr",
        "appOwnerMnemonic": "call brass keen rough true spy dream robot useless ignore anxiety badge",
        "appOwnerWriteKey": {
          "data": [
            193,
            200,
            155,
            104,
            173,
            156,
            172,
            55,
            27,
            230,
            228,
            51,
            167,
            204,
            43,
            8,
          ],
          "type": "Buffer",
        },
        "clock": "1970-01-01T00:00:00.000Z-0001-7b30e5c418ca52c3",
      },
    ],
  },
  {
    "name": "evolu_history",
    "rows": [
      {
        "column": "name",
        "id": {
          "data": [
            142,
            234,
            71,
            9,
            213,
            68,
            164,
            29,
            160,
            190,
            95,
            223,
            168,
            17,
            36,
            152,
          ],
          "type": "Buffer",
        },
        "ownerId": {
          "data": [
            7,
            80,
            102,
            185,
            48,
            114,
            212,
            210,
            120,
            45,
            146,
            214,
            157,
            250,
            36,
            172,
          ],
          "type": "Buffer",
        },
        "table": "testTable",
        "timestamp": {
          "data": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            1,
            123,
            48,
            229,
            196,
            24,
            202,
            82,
            195,
          ],
          "type": "Buffer",
        },
        "value": "test",
      },
    ],
  },
  {
    "name": "testTable",
    "rows": [
      {
        "createdAt": null,
        "id": "jupHCdVEpB2gvl_fqBEkm",
        "isDeleted": null,
        "name": "test",
        "updatedAt": "1970-01-01T00:00:00.000Z",
      },
    ],
  },
  {
    "name": "_testTable",
    "rows": [],
  },
  {
    "name": "evolu_timestamp",
    "rows": [
      {
        "c": 1,
        "h1": 112010336006712,
        "h2": 246093702771481,
        "l": 2,
        "ownerId": {
          "data": [
            7,
            80,
            102,
            185,
            48,
            114,
            212,
            210,
            120,
            45,
            146,
            214,
            157,
            250,
            36,
            172,
          ],
          "type": "Buffer",
        },
        "t": {
          "data": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            1,
            123,
            48,
            229,
            196,
            24,
            202,
            82,
            195,
          ],
          "type": "Buffer",
        },
      },
    ],
  },
]
`;

exports[`reset 1`] = `
{
  "schema": {
    "indexes": [],
    "tables": [],
  },
  "tables": [],
}
`;
